### ISSUE TITLE
Support `asChild` like in radix-ui

### ISSUE THREAD

**Issue by latin-1** [ariakit/ariakit#1245]
### Motivation

Currently, we have two ways to composite components: 1. use `as` prop or 2. use `children` as render props. But both of them have their limitations, for example, a button with a `Tooltip`:

1. `as` prop: it requires a little bit more code changes to add/remove a tooltip on an element, and it cannot composite more than two components, e.g., A as B as C.

   ```diff
   - <Button onClick={onClick}>
   + <TooltipAnchor as={Button} onClick={onClick}>
   ```
2. `children` as render-props: it requires us to merge the props ourselves, which seems weird.

   ```ts
   <TooltipAnchor>
     {props => <Button {mergeProps({ onClick }, props)} />}
   ```

We suggest a new way to composite components by introducing a new prop, `asChild`.

```ts
<TooltipAnchor asChild>
  <Button onClick={onClick}>
```

Internally, it will `cloneElement(children)` with merged props, like `as` prop, but more intuitive.

See: https://www.radix-ui.com/docs/primitives/overview/styling#changing-the-rendered-element

**Comment by diegohaz** [comment]
> 1. `as` prop: it requires a little bit more code changes to add/remove a tooltip on an element

[There's a lot of problems with the `as` prop](https://youtu.be/_kyv3tnQKfg?t=949), but I don't understand why that would be one, especially comparing to `asChild`:

```diff
- <Button onClick={onClick}>
+ <TooltipAnchor asChild>
+   <Button onClick={onClick} />
```

> 2\. `children` as render-props: it requires us to merge the props ourselves.

You don't need to merge the props yourself. Just pass them to `TooltipAnchor`, which will merge them internally.

```jsx
<TooltipAnchor onClick={onClick}>
  {props => <Button {...props} />}
```

But this API also allows you to control how the props will be merged. For example, the Ariakit components will always call your event handlers before performing any internal logic. But you may want to run your click handler *after* the internal one. You're free to come up with your own `mergeProps` function.

You can also do that with `asChild`, but you need to create a separate component, making it less convenient. The `asChild` and `as` props suffer from the same problems when people need to create components. People will often forget to spread the props and/or forward refs.

`asChild` has its own tradeoffs. Because of its implicit magic, you don't get some of the benefits of having direct access to the props.

For example, if you're using something like Tailwind and you want to style the active menu item, you can do this:

```jsx
<MenuItem>
  {(props) => {
    const className = "data-active-item" in props ? "bg-blue" : "";
    return <div {...props} className={className} />;
  }}
</MenuItem>
```

On the website, we're using the `hidden` prop from `TabPanel` to show the panel contents conditionally:

https://github.com/ariakit/ariakit/blob/8f42d33f02f9c782b4096d43c959ffea837dceb7/packages/website/components/playground/index.tsx#L216-L225

With render props, you can pass the props to a different prop on the child component:

```jsx
<MenuItem>
  {(props) => <MyCustomItem {...props} innerRef={props.ref} />}
</MenuItem>
```

You can render other elements:

```jsx
<Radio>
  {(props) => (
    <>
      <input {...props} />
      <label htmlFor={props.id}>Label</label>
    </>
  )}
</Radio>
```

[Render props also have some problems](https://youtu.be/_kyv3tnQKfg?t=606), but, as far as I can see, the `asChild` prop doesn't solve any of them, and has more limitations.

I'm not saying we shouldn't support `asChild`. It does solve some of the problems with the `as` prop (more specifically, TypeScript and prop collisions). But so does render props. And I would be wary to add yet another API to solve problems that the existing APIs already solve.

Anyway, you can always create your own abstractions on top of Ariakit components that support the `asChild` prop.

**Comment by jjenzz** [comment]
always great to read other perspectives on this as radix has historically had requests for render-prop support that have been rejected, because everything described here is also possible with `asChild`. 

the preference was for it to force the consumer to **opt-in** to an increased boilerplate/less performant render prop approach instead of consumers potentially having to reverse engineer perf issues later.

https://codesandbox.io/s/optimistic-einstein-3qytt4?file=/src/App.js

> You can also do that with `asChild`, but you need to create a separate component, making it less convenient.

i'm curious why you believe that creating one re-usable component like the above is "less convenient" than having to:

- type out render prop boilerplate instead of auto-merge which is often what ppl want
- consider perf implications of render prop
- lose ability to intuitively `React.memo`

ofc, a lot of this comes down to personal preference but with `asChild` you can easily chose either/or without hindering perf for more junior devs. 

you cannot easily opt-in to an `asChild` approach from a render-prop base without having to carefully consider the above issues.

> For example, the Ariakit components will always call your event handlers before performing any internal logic. But you may want to run your click handler after the internal one.

i generally believe that to avoid unexpected side-effects, consumers should avoid re-ordering event handlers as it will prevent browser APIs like `event.preventDefault()` from behaving as expected. this again is why radix preference is for more experienced devs to create a component (an assumption that they know what they're doing) so that more junior devs have everything working as the DOM would, out of the box.

> `asChild` has its own tradeoffs. Because of its implicit magic, you don't get some of the benefits of having direct access to the props.
> 
> For example, if you're using something like Tailwind and you want to style the active menu item, you can do this:
> 
> ```jsx
> <MenuItem>
>  {(props) => {
>    const className = "data-active-item" in props ? "bg-blue" : "";
>    return <div {...props} className={className} />;
>   }}
> </MenuItem>
> ```
>

the `asChild` prop isn't needed for that and I would personally suggest against render-prop there too because it is forcing the consumer to redeclare the component semantics (DOM structure) by specifying a `div` to put the `className` onto. instead, radix provides a controlled api for things like this so that there is no need to redeclare semantics:

```jsx
const [active, setActive] = React.useState('one');

<Menu value={active} onValueChange={setActive}>
  <MenuItem value="two" className={active === 'two' ? 'bg-blue': undefined} />
</Menu>
```

> [Render props also have some problems](https://youtu.be/_kyv3tnQKfg?t=606), but, as far as I can see, the asChild prop doesn't solve any of them, and has more limitations.

i'm curious which of these issues you believe `asChild` doesn't solve. AFAIK, it solves them all.

**Comment by latin-1** [comment]
@diegohaz 

> [There's a lot of problems with the as prop](https://youtu.be/_kyv3tnQKfg?t=949), but I don't understand why that would be one, especially comparing to asChild:

It's not only about git-diff stuff, but also the developer's effort to add/remove a tooltip. For `asChild`, we only need to wrap/unwrap the button by a `TooltipAnchor`, and leave the button itself as-is. BTW personally, I ignore whitespaces while code-reviewing so that the diff would look like this:

```diff
+<TooltipAnchor asChild>
   <Button onClick={onClick}>
```

> But this API also allows you to control how the props will be merged. For example, the Ariakit components will always call your event handlers before performing any internal logic. But you may want to run your click handler after the internal one. You're free to come up with your own mergeProps function.

I don't think this is necessary. Assume you have a `<a>` and want to add a listener after the navigation happens. You might use `requestAnimationFrame` to do so. The same happens to Ariakit, too.

> For example, if you're using something like Tailwind and you want to style the active menu item, you can do this:

Checking the props that passed to the render props seems weird. I'd prefer to get it from the `state` directly. That's what `state` is designed to be.

**Comment by latin-1** [comment]
@diegohaz 

> You don't need to merge the props yourself. Just pass them to TooltipAnchor, which will merge them internally.

It's not type-safe either.

```ts
<TooltipAnchor variant="primary">
               ^ type error
  {(props) => (
    <Button {...props}>
```

Also, `children` cannot be hoisted to the `Tooltip`. It requires a policy to split props into two parts in order to build abstractions on Ariakit.

**Comment by diegohaz** [comment]
@jjenzz Thanks for chiming in. Your thoughts are greatly appreciated.

> i'm curious why you believe that creating one re-usable component like the above is "less convenient"

I'm sorry if I'm being too obvious; my logic is that having a render prop API is more convenient than creating an adapter if you *do need* to fall back to render props.

Moreover, suppose people don't fall back to render props and instead create an intermediate component. In that case, it's too easy to forget to spread props because you're not explicitly passing them to the intermediate component. The `as` prop has the same problem. But I haven't seen this happening with render props (people still forget to `forwardRef` though, but I'd say this is more of React's fault).

> * consider perf implications of render prop

What are the perf implications?

> * lose ability to intuitively `React.memo`

I'm curious about this one. I remember doing a lot of experiments with `React.cloneElement` a while ago, and the incompatibility with `React.memo` was one of the biggest issues.

```jsx
<AsChild>
  <MemoComponent /> {/* React.memo is always invalidated */}
</AsChild>

<RenderProp>
  {(props) => <MemoComponent {...props} />} // React.memo works
</RenderProp>
```

I put together this CodeSandbox to demonstrate the issue: https://codesandbox.io/s/react-memo-with-aschild-and-render-props-vwzm56?file=/src/App.js

Let me know if I'm missing something.

**Comment by diegohaz** [comment]
@latin-1 

> I don't think this is necessary. Assume you have a `<a>` and want to add a listener after the navigation happens. You might use `requestAnimationFrame` to do so. The same happens to Ariakit, too.

That's not the same. My suggestion keeps the logic synchronous. `requestAnimationFrame` will happen asynchronously, and other events may be triggered during this time.

> Checking the props that passed to the render props seems a little bit weird. I'd prefer to get it from the `state` directly. That's what `state` is designed to be.

If you have access to the state, yes. But the Ariakit (v2) API allows you to abstract the child components without the state:

```jsx
const MyMenuItem = forwardRef((props, ref) => {
  return (
    <MenuItem {...props} ref={ref}>
      {(props) => {
        const className = "data-active-item" in props ? "bg-blue" : "";
        return <div {...props} className={className} />;
      }}
    </MenuItem>
  );
});
```

You can always pass the `state.activeId` value as a prop to your component, but that's more work, especially if you want to `React.memo` your component.

> It's not type-safe either.
> 
> ```ts
> <TooltipAnchor variant="primary">
>                ^ type error
>   {(props) => (
>     <Button {...props}>
> ```

If `variant` is a specific prop of `Button`, you should pass it to `Button`. TypeScript will not complain if you just spread `{...props}` onto `TooltipAnchor` and it includes `variant`, though.

---

Regarding creating an `asChild` API with the current Ariakit API, you can do it with [render props](https://codesandbox.io/s/ariakit-aschild-qbfucf?file=/src/tooltip.tsx) or by creating a reusable `AsChild` component and passing it to the Ariakit `as` prop. You can also use Radix Slot, by the way: https://codesandbox.io/s/ariakit-radix-slot-tgi9er?file=/src/tooltip.tsx

**Comment by latin-1** [comment]
@diegohaz 

> I'm curious about this one. I remember doing a lot of experiments with React.cloneElement a while ago, and the incompatibility with React.memo was one of the biggest issues.

I believe there's nothing wrong with `cloneElement`. Many things could make `memo` not work, event handlers without `useCallback`, children without `useMemo`, etc. Both render props and `asChild` should work fine — if you take care of all props, including children, which is almost impossible.

> If variant is a specific prop of Button, you should pass it to Button. TypeScript will not complain if you just spread {...props} onto TooltipAnchor and it includes variant, though.

The problem is that we need to pick props that need to hoist carefully to avoid potential issues. In my opinion, Ariakit should work as a black box. Developers should not care about the internal mechanism.

> Regarding creating an asChild API with the current Ariakit API, you can do it with [render props](https://codesandbox.io/s/ariakit-aschild-qbfucf?file=/src/tooltip.tsx) or by creating a reusable AsChild component and passing it to the Ariakit as prop. You can also use Radix Slot, by the way: https://codesandbox.io/s/ariakit-radix-slot-tgi9er?file=/src/tooltip.tsx

`Slot` component (either from Radix or build a custom one) will use a non-optimized `mergeProps`, which is less performant. To get the full benefit of Ariakit, we still need to do some dirty work. Unlike in the user space, implementing `asChild` in Ariakit will be more straightforward.

**Comment by diegohaz** [comment]
> I believe there's nothing wrong with `cloneElement`. Many things could break `memo`, event handlers without `useCallback`, children without `useMemo`, etc. Both render props and `asChild` should work fine — if you take care of all props, including children, which is almost impossible.

That's true. I guess I wasn't memoizing the `ref` prop back when I experimented with it. Now I only tested with Radix's Slot assuming it was optimized for it.

> In my opinion, `memo` on UI components like button doesn't make sense. See also [Before You memo()](https://overreacted.io/before-you-memo/).

Yes, that was a contrived example.

> The problem is that we need to pick props that need to hoist carefully to avoid potential issues. In my opinion, Ariakit should work as a black box. Developers should not care about the internal mechanism.

I'm not sure I understand your point. What are the potential issues you're talking about? How does my suggestion interact with internal mechanisms?

**Comment by jjenzz** [comment]
> Moreover, suppose people don't fall back to render props and instead create an intermediate component. In that case, it's too easy to forget to spread props because you're not explicitly passing them to the intermediate component. The `as` prop has the same problem.

i agree, this is a **huge** pain, but like you say the `as` prop has the same issue and yet people have flocked to use it which tells me they prefer the convenience of that kind of api regardless of this pitfall. 

it seems the benefit outweighs the cost for most here because we have to make the intermediate component accept those props regardless. with `asChild` you can update the intermediate component declaration once and forget about it (like `React.forwardRef`) and now it works anywhere without any additional render-prop overhead. 

> What are the perf implications?
>  I remember doing a lot of experiments with `React.cloneElement` a while ago, and the incompatibility with `React.memo` was one of the biggest issues.

i believe you touched on it in your video that render-prop will mean a new `children` reference every render but i hadn't honestly ever considered that `React.cloneElement` had these issues too 🤯  my bad.

for radix, these perf issues were not a motivator though because radix team is pretty against `React.memo`/premature optimisation. they were an added bonus (we thought 😅 ) but we chose `asChild` due to it being more frictionless, especially when more junior devs need to create abstractions like the following:

```jsx
const Tooltip = ({ children, content }) => {
  return (
    <Tooltip.Root>
      <Tooltip.Trigger asChild>{children}</Tooltip.Trigger>
      <Tooltip.Content>{content}</Tooltip.Content>
    </Tooltip.Root>
  );
};
```

vs: 

```jsx
import mergeProps from 'merge-props';

const Tooltip = ({ children, content }) => {
  return (
    <Tooltip.Root>
      <Tooltip.Trigger>
        {(props) => {
          const child = React.Children.only(children);
          // make sure props are merged in the correct order (not sure if i have here)
          return React.cloneElement(child, mergeProps(child.props, props));
        }}
      </Tooltip.Trigger>
      <Tooltip.Content>{content}</Tooltip.Content>
    </Tooltip.Root>
  );
};
```

having said all of the above tho, i don't think `asChild` and render-props are interchangeable. i view `asChild` as a way to replace the underlying node while maintaining functionality and render-props as a tool for reaching into internal state. 

render-props vs controlled api would be a more apt. comparison imo so my points above are about `asChild` **in the absence of an `as` prop** because the `asChild` prop was created to replace `as` only. if ariakit has an `as` prop then i don't see why it would need `asChild` as well. 

> `Slot` component (either from Radix or build a custom one) will use a non-optimized mergeProps, which is less performant.

@latin-1 it would be great if you could raise an issue for the `Slot` component on the radix repo outlining your concerns here because i'm not entirely sure what the issues are that you're describing.

**Comment by diegohaz** [comment]
I'm thinking about supporting this feature, but with the name `forwardProps` or `passProps` to make it more obvious that the child will receive the props (and therefore should handle them). What do you think?

**Comment by latin-1** [comment]
I'm glad to hear that! I don't have a strong opinion about naming. Both are acceptable to me.

**Comment by jasikpark** [comment]
++ this feature would solve my issue using `<MenuItem as={ReactRouterV6Link} state={{ from: location }}>lol</MenuItem>`, since the components have [conflicting `state` props](https://reactrouter.com/en/main/components/link#state).

I'm a big fan of the `element={<Element/>}` pattern used in Suspense, ErrorBoundaries, and now `react-router-dom@v6`

**Comment by diegohaz** [comment]
@jasikpark You can use render props. See https://github.com/ariakit/ariakit/discussions/2278#discussioncomment-5643241

**Comment by jasikpark** [comment]
ah, ty!
